File: cmd/afk.py
import time
import discord
from discord.ext import commands
from discord import app_commands

from utils.embeds import make_embed
from db.db_helpers.afk import set_afk


class AFK(commands.Cog):

    def __init__(self, bot: commands.Bot):
        self.bot = bot

    @app_commands.command(name="afk",
                          description="Mark yourself as AFK in this server")
    @app_commands.describe(reason="Reason for going AFK (optional)")
    async def afk(self, interaction: discord.Interaction, reason: str = "AFK"):
        # Slash commands can technically be used in DMs
        if not interaction.guild:
            embed = make_embed(
                title="Invalid Context",
                description="AFK status can only be set inside a server.",
                level="ERROR")
            return await interaction.response.send_message(embed=embed,
                                                           ephemeral=True)

        now = int(time.time())

        set_afk(guild_id=interaction.guild.id,
                user_id=interaction.user.id,
                reason=reason)

        embed = make_embed(title="AFK Enabled",
                           description=("You are now marked as AFK.\n"
                                        f"Reason: {reason}\n"
                                        f"Since: <t:{now}:R>"),
                           level="SUCCESS",
                           footer=f"Server: {interaction.guild.name}")

        await interaction.response.send_message(embed=embed, ephemeral=True)


async def setup(bot: commands.Bot):
    await bot.add_cog(AFK(bot))

File: src/commands/utility/afk.js
import { SlashCommandBuilder } from "discord.js";
import { createCommand } from "../../core/command.js";
import { makeEmbed } from "../../core/embeds.js";
import { setAfkStatus } from "../../db/helpers/afk.js";

const slashBuilder = new SlashCommandBuilder()
  .setName("afk")
  .setDescription("Set your Away-From-Keyboard status for this server")
  .addStringOption((opt) => opt.setName("reason").setDescription("Why are you going AFK?").setRequired(false));

export default createCommand({
  name: "afk",
  description: "Set your Away-From-Keyboard status for this server",
  category: "Utility",
  slashBuilder,

  async execute(ctx) {
    if (!ctx.guild) {
      return await ctx.reply({
        content: "The AFK command can only be used inside a server.",
        ephemeral: true,
      });
    }

    const reason = ctx.options.reason || ctx.options.raw || "AFK";

    let originalNick = ctx.member?.nickname || null;
    if (
      ctx.member &&
      ctx.guild?.members?.me?.permissions?.has("ManageNicknames") &&
      !ctx.member.permissions.has("Administrator")
    ) {
      const newNick = `[AFK] ${ctx.member.displayName}`.slice(0, 32);
      await ctx.member.setNickname(newNick).catch(() => {});
    }

    await setAfkStatus(ctx.user.id, ctx.guild.id, reason, originalNick);

    const avatar = ctx.user.displayAvatarURL({ dynamic: true, size: 256 });
    const embed = makeEmbed({
      author: {
        name: "Away From Keyboard",
        iconURL: avatar,
      },
      title: `${ctx.user.username} is now AFK`,
      description:
        `• **Reason:** \`${reason}\`\n\n` +
        `-# Sending a message in this server will clear your AFK status and show any mentions you received.`,
      thumbnail: avatar,
      level: "SUCCESS",
      headerDivider: false,
    });

    return await ctx.reply({ embeds: [embed] });
  },
});
